# Working with API Requests

Working with API requests involves more than just making simple HTTP calls. It requires knowledge of various concepts and techniques to handle different scenarios, such as handling response codes, working with headers, dealing with authentication, understanding CORS restrictions, parsing API responses, implementing pagination, handling errors, and following best practices like API versioning. We will learn more about these advanced concepts in this lesson.

## HTTP headers

Headers provide important information to the server about the request being made. They convey instructions, preferences, or data related to the request, allowing the server to handle the request appropriately.

### Adding Headers to API Requests

To add headers to an API request, you typically include them in the HTTP header section of the request. The specific method for adding headers depends on the programming language or library you are using. Here's an example using the Python `requests` library:


In [ ]:
import requests

url = "https://api.example.com/endpoint"
headers = {
    "Content-Type": "application/json",
    "Authorization": "Bearer your_token_here"
}

response = requests.get(url, headers=headers)

In this example, we define a `headers` dictionary containing the desired headers. The `"Content-Type"` header specifies that the request body is in JSON format, while the `"Authorization"` header includes an authentication token using the Bearer token scheme. We will discuss more about different types of header and authentication token in a latter section

We pass the `headers` dictionary to the `requests.get()` method as the headers parameter.

### Common Headers

#### `Content-Type`

The `"Content-Type"` header specifies the media type of the data in the request or response body. It informs the server or client about how to interpret and handle the data. Common values for the `"Content-Type"` header include:
- `application/json`: This indicates that the request or response body contains data in JSON format. JSON (JavaScript Object Notation) is widely used for data serialization and exchange in web APIs.
- `application/xml`: This signifies that the data is in XML format. XML (eXtensible Markup Language) is another popular format for representing structured data.
- `application/x-www-form-urlencoded`: This is commonly used for HTML form submissions. Data in this format is URL-encoded, where key-value pairs are encoded and joined with ampersands (`&`).

#### `Authorization`

The `"Authorization"` header is crucial for authentication and authorization purposes. It carries credentials or tokens to prove the identity of the client making the request. Some common examples of the `"Authorization"` header include:
- `Bearer Token`: In the bearer token scheme, the token is included as follows: `Authorization: Bearer your_token_here`. Bearer tokens are often used in token-based authentication systems, where a client obtains a token and includes it in subsequent API requests to access protected resources.
- `Basic Authentication`: Basic authentication involves encoding the username and password in Base64 and including them in the header as follows: `Authorization: Basic base64_encoded_credentials`. Basic authentication is a simple mechanism for sending credentials, but it is less secure compared to token-based authentication.

#### `Accept`

The `"Accept"` header allows the client to specify the media types it can accept in the response. It informs the server about the preferred format of the response data. For example: `Accept: application/json`.

By indicating the desired format, the client can communicate its expectations to the server, which can then provide the response in the appropriate format.

#### `User-Agent`

The `"User-Agent"` header in an API request identifies the client application or user agent that is making the request. It provides information about the software, device, and platform used by the client. The `User-Agent` header is included in the HTTP request headers and is typically sent by web browsers, mobile apps, or other HTTP clients.

For example: `User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36`.

In this example, the `User-Agent` indicates that the request is coming from a web browser running on Windows 10 (Windows NT 10.0). The browser is Google Chrome version 91.0.4472.124, and it utilizes the WebKit rendering engine.

## Authentication Tokens

Authentication in APIs involves verifying the identity of a client, such as a user or an application, before granting access to protected resources. It ensures that only authorized individuals or entities can interact with the API and perform specific actions. Authentication typically occurs during the initial stages of an API request, where the client provides credentials or tokens to prove its identity.

The primary objectives of authentication in APIs are:
- **Identity Verification**: Ensuring that the client's claimed identity is valid and can be trusted
- **Access Control**: Granting or denying access to specific resources based on the client's authenticated identity and authorization privileges

### Authentication Token Concepts

Authentication tokens are commonly used in APIs to authenticate and authorize client requests. Here are a few key concepts related to authentication tokens:

#### 1. API Keys

*API keys* are unique identifiers issued to clients (e.g., applications) by the API provider. They act as a form of credentials that clients include in API requests to authenticate themselves. API keys are typically long alphanumeric strings that can be used to uniquely identify the client making the request. Clients include the API key in the request headers, such as the `"X-API-Key"` header, to authenticate themselves.

#### 2. Access Tokens

*Access tokens* are temporary credentials issued to authenticated clients after a successful authentication process. Access tokens are used to authorize subsequent API requests. They are typically short-lived and have an expiration time. Access tokens can be more secure than API keys as they have a limited validity period, reducing the risk if they are compromised.

#### 3. Refresh Tokens

Refresh tokens are additional tokens issued alongside access tokens. They are used to obtain new access tokens when the original access token expires. Refresh tokens are usually long-lived and can be used to maintain an authenticated session without requiring the client to re-enter their credentials. Refresh tokens need to be securely stored and transmitted since they grant the ability to obtain new access tokens.

### Different Types of Authentication Mechanisms

Several authentication mechanisms are commonly used in APIs. Here are a few examples:

- **Basic Authentication**: This mechanism involves sending the client's credentials (e.g., username and password) encoded in Base64 format in the `"Authorization"` header of the API request. Basic authentication is simple to implement but may be less secure as credentials are sent with each request.

- **Bearer Token Authentication**: Bearer token authentication relies on including an access token in the `"Authorization"` header of the API request using the Bearer token scheme. The access token represents the client's authentication and authorization for accessing protected resources.

- **OAuth 2.0**: *OAuth 2.0* is an authorization framework that allows third-party applications to access resources on behalf of a user. It involves exchanging authorization codes, access tokens, and refresh tokens to grant access to protected resources while maintaining user privacy and security.

- **JSON Web Tokens (JWT)**: *JWT* is a compact, self-contained format for securely transmitting information between parties as a JSON object. JWTs can be used as access tokens to authenticate and authorize API requests. They are digitally signed to ensure integrity and can carry additional claims about the client.

> These are just a few examples of authentication mechanisms used in APIs. The choice of authentication mechanism depends on factors such as security requirements, client capabilities, and the API provider's preferences.

## Including Authentication Tokens in API Requests

To include an authentication token in the headers of a `requests` library API request, you can pass a dictionary containing the headers using the `headers` parameter. Here's an example:

In [ ]:
import requests

url = 'https://api.example.com/resource'
headers = {'Authorization': 'Bearer YOUR_ACCESS_TOKEN'}

response = requests.get(url, headers=headers)

In the above example, the headers dictionary includes the `"Authorization"` key with a value of `"Bearer YOUR_ACCESS_TOKEN"`. This adds the authentication token to the request headers.

### Including Tokens as Query Parameters

To include an authentication token as a query parameter in the URL, you can pass the token as part of the URL string. Here's an example:

In [ ]:
import requests

url = 'https://api.example.com/resource?access_token=YOUR_ACCESS_TOKEN'

response = requests.get(url)

In this example, the authentication token `"YOUR_ACCESS_TOKEN"` is appended to the URL as the `access_token `query parameter.

Both approaches can be used with different HTTP methods (`GET`, `POST`, `PUT`, etc.) and are not limited to just retrieving resources (`GET` requests). You can modify the examples based on your specific use case and the requirements of the API you are interacting with.

> Remember to replace `"YOUR_ACCESS_TOKEN"` with the actual value of your authentication token when using these examples. Additionally, ensure that you are following the guidelines and requirements provided by the API documentation or authentication system to include the token correctly.

## Authentication Examples

To illustrate an example, let's consider the Github API. Some requests to this API can be made without authentication. For instance, you can retrieve a list of repositories belonging to a user by sending the following request:

In [3]:
import requests

response = requests.get('https://api.github.com/users/IvanYingX/repos')
repos = response.json()

for repo in repos:
    print(repo['name'])


Actions
actions_aicore_test
AiBetCore
AiCore-Project-Template
aicore-teachable-machine
AiCore-Test2
angular-clipboard-auto
animal_scraper
aws-elastic-beanstalk-express-js-sample
cat_scraper
Challenges_AiCore
colab_test
computer-vision-rock-paper-scissors
Content-Review-Action
Coursera_Capstone
Daily_Checklist
daneah
data-collection-pipeline
Data-Engineering-Project
Data-Science-Project
Data_Science_Template
Debugging_practical
Deploy-Content-Action
didactic-umbrella
dlaicourse
dvc_test
Essentials-Project
ExampleDocumentation
Facebook-Demo
Face_Mask_RPi



In this example, we utilize the `requests` library in Python to make an HTTP `GET` request to the specified API endpoint. The response is then parsed as JSON, allowing us to access the list of repositories. Finally, we iterate over the repositories and print out their names.

However, certain actions, such as creating a repository, require authentication. To create a repository, you need to send a `POST` request to the API with the required data. Here's an example using the `requests` library:

In [4]:
import json

response = requests.post('https://api.github.com/user/repos', data=json.dumps({"name": "Repo-From-Python"}))
print(response.json)

<bound method Response.json of <Response [401]>>


The error code `401` signifies an `"Unauthorized"` error, which means the request lacks valid authentication credentials. In the case of the Github API, you need to provide proper authentication to create a repository.

To successfully authenticate, we need to generate a token from Github and add this token to the header of the request. You can create a token in the Github developer portal. First, go to your profile and click on **Settings**.

<p align="center">
    <img src="images/github_setting.png" width="400" height="600"/>
</p>

Then click on **Developer Settings**.

<p align="center">
    <img src="images/Dev_setting.png" width="700" height="500"/>
</p>

Finally in **Personal access tokens** click **Generate new token**.

<p align="center">
    <img src="images/PAT.png" width="700" height="300"/>
</p>

> Copy the generated token and make a note of it on your local machine. You will have to use this token in your request.

In [ ]:
import requests
import json

# API endpoint for repository creation
url = 'https://api.github.com/user/repos'

# Authentication token
access_token = 'YOUR_TOKEN_HERE' # replace with your own token if you are following the walkthrough

# Repository data
repo_data = {
    'name': 'Repo-From-Python',
    'description': 'This is a new repository created via the API',
    'private': False
}

# Set the headers with the authentication token
headers = {
    'Authorization': f'token {access_token}',
    'Accept': 'application/vnd.github.v3+json'
}

# Send the POST request to create the repository
response = requests.post(url, data=json.dumps(repo_data), headers=headers)

print('Response status code:', response.status_code)


A status code of `201` means that the request was successful. When you run the code you will get a `401` status code, as this token has been used only for demonstration purposes and has since been removed. Let's checkout the newly created repo:

<p align="center">
    <img src="images/repo.png" width="700" height="500"/>
</p>

## Cross-Origin Resource Sharing (CORS)

> *Cross-Origin Resource Sharing (CORS)* is a mechanism that allows resources (e.g., APIs) on a web page to be requested from another domain outside the domain from which the resource originated. 

The primary purpose of CORS in web APIs is to protect user data and prevent unauthorized access by enforcing strict controls on cross-origin requests. Without CORS, malicious websites could make cross-origin requests to sensitive APIs and retrieve or manipulate data without the user's knowledge or consent. CORS adds an additional layer of security by imposing restrictions on such requests, allowing only trusted origins to access resources.

> By default, web browsers enforce the same-origin policy, which means that web pages can only make requests to resources within the same origin (protocol, domain, and port combination) as the web page itself. CORS relaxes this policy and allows controlled cross-origin requests while maintaining security.

### How Does CORS Work?

CORS involves a series of requests and responses between the client-side application and the server to establish trust and enable cross-origin communication. The key components of CORS are:

- *Origin*: The combination of the protocol, domain, and port from which a web page is served. For example, https://www.example.com.
- *Cross-Origin Request*: A request sent by a web page from one origin to a resource on a different origin
- *Preflight Request*: An HTTP OPTIONS request sent by the browser to the server before making certain types of cross-origin requests. The preflight request verifies whether the actual request is safe to send by checking the server's CORS policy.
- *CORS Headers*: Special headers exchanged between the client and server to negotiate and enforce the CORS policy. The key CORS headers include `Access-Control-Allow-Origin`, `Access-Control-Allow-Methods`, `Access-Control-Allow-Headers`, and `Access-Control-Allow-Credentials`.

### Benefits of CORS

CORS provides several benefits for web developers and API providers:
- **Secure Cross-Origin Communication**: CORS allows controlled access to resources from different origins, ensuring secure communication between web applications
- **Enhanced Web Application Functionality**: CORS enables web applications to utilize and consume resources from external APIs, enhancing their functionality and enabling integration with third-party services
- **Standardized Approach**: CORS provides a standardized mechanism for handling cross-origin requests, making it easier to implement and understand
- **Granular Control**: Servers can specify precise CORS policies, allowing or restricting access to specific origins, methods, or headers

## Key Takeaways

- HTTP headers play a crucial role in API requests and responses. They provide additional information, such as content type and authorization, that allows clients and servers to communicate effectively.
- Common headers, such as `Content-Type` and `Authorization`, have specific purposes and usage in API requests. Understanding their significance helps ensure proper communication with APIs.
- Authentication tokens are essential for securing API requests. API keys and access tokens are commonly used authentication mechanisms. They grant authorized access to resources and help ensure the integrity of data exchanged between clients and servers.
- CORS (Cross-Origin Resource Sharing) is a mechanism that enables controlled cross-origin communication in web APIs